### PR TITLE
Rd 11070 add amqplib support

### PR DESCRIPTION
**Pull request by therightstuff**

### CODE REVIEW

File: src/instrumentations/amqplib/AmqplibInstrumentation.test.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    jest.clearAllMocks();
+  });
+
+  let lumigoAmqplibInstrumentation = new LumigoAmqplibInstrumentation();
+
+  test('getInstrumentedModule should return "amqplib"', () => {
+    expect(lumigoAmqplibInstrumentation.getInstrumentedModule()).toEqual('amqplib');
+  });
+
+  test.skip('requireIfAvailable should return required name', () => {
```
Comment (saartochner) [inline src/instrumentations/amqplib/AmqplibInstrumentation.test.js:14]: why did you add skipped test? 🤔

File: src/instrumentations/amqplib/AmqplibInstrumentation.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+export default class LumigoAmqplibInstrumentation extends Instrumentor<AmqplibInstrumentation> {
+  getInstrumentedModule(): string {
+    return 'amqplib';
+  }
+
+  getInstrumentation(): AmqplibInstrumentation {
+    return new AmqplibInstrumentation({
+      publishHook: (span: Span, publishInfo: PublishInfo) => {
+        span.setAttribute('messaging.publish.body', publishInfo.content.toString());
```
Comment (saartochner) [inline src/instrumentations/amqplib/AmqplibInstrumentation.ts:17]: can you limit it?
use the util `add_body_attribute`

File: src/instrumentations/amqplib/AmqplibInstrumentation.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      },
+      consumeHook: (span: Span, consumeInfo: ConsumeInfo) => {
+        span.setAttribute('messaging.consume.body', consumeInfo.msg.content.toString());
```
Comment (saartochner) [inline src/instrumentations/amqplib/AmqplibInstrumentation.ts:20]: here too

File: test/instrumentations/amqplib/amqplib.test.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+       * This prevents spurious failures of early tests.
+       */
+      try {
+        rabbitmqContainer = await startRabbitMqContainer();
+      } finally {
+        if (rabbitmqContainer) {
+          rabbitmqContainer.stop();
+        }
+      }
+    }, STARTUP_TIMEOUT /* Long timeout, this might have to pull Docker images */);
```
Comment (saartochner) [inline test/instrumentations/amqplib/amqplib.test.ts:56]: maybe it's worth adding a print before it?
So when you have slow network you'll know that this is the reason

File: test/instrumentations/amqplib/amqplib.test.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+        let spans = await testApp.invokeGetPathAndRetrieveSpanDump(
+          `/invoke-amqp-consumer?topic=${topic}&message=${message}&host=${host}&port=${port}`
+        );
+
+        /*
+         * TODO: HORRIBLE WORKAROUND: The internal span we are looking for seems to be closed LATER than
+         * the Server span, so we must busy-poll.
+         */
+        while (spans.length < 4) {
```
Comment (saartochner) [inline test/instrumentations/amqplib/amqplib.test.ts:107]: an idea - I see this "horrible workaround" everywhere.
WDYT to extract it to a single util function, and we will fix it once [if] it will be possible?
There's a lot of code duplication around here

File: test/instrumentations/amqplib/amqplibTestUtils.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+}) {
+  let messageKey = 'invalid-message-key';
+  switch (spanKind) {
+    case SpanKind.PRODUCER:
+      messageKey = 'messaging.publish.body';
+      break;
+    case SpanKind.CONSUMER:
+      messageKey = 'messaging.consume.body';
+      break;
+  }
```
Comment (saartochner) [inline test/instrumentations/amqplib/amqplibTestUtils.js:36]: let's try to write more explicit code:
```suggestion
 if spanKind ==  SpanKind.PRODUCER
         messageKey = 'messaging.publish.body';
 else if spanKind == SpanKind.CONSUMER
        messageKey = 'messaging.consume.body';
 else
       assert(false);
```

**Review by saartochner — APPROVED** [review]
great work! approved in advanced, please fix the comments

File: src/instrumentations/amqplib/AmqplibInstrumentation.test.js (hunk 1 above)
Comment (therightstuff) [reply src/instrumentations/amqplib/AmqplibInstrumentation.test.js:14]: i didn't intend to...

Comment (therightstuff) [reply src/instrumentations/amqplib/AmqplibInstrumentation.test.js:14]: i take that back, it wasn't a mistake, it's a copy paste from the other instrumentation tests. i'm going to unskip it and see what happens

Comment (therightstuff) [reply src/instrumentations/amqplib/AmqplibInstrumentation.test.js:14]: it's not clear why these tests were skipped, so i'm unskipping the lot of them (incl. for grpc)

File: src/instrumentations/amqplib/AmqplibInstrumentation.ts (hunk 2 above)
Comment (therightstuff) [reply src/instrumentations/amqplib/AmqplibInstrumentation.ts:17]: @saartochner did you mean use something like the `concatenatePayload` method you implemented for grpc? should that not be a general utility method for all the instrumentations?

Comment (saartochner) [reply src/instrumentations/amqplib/AmqplibInstrumentation.ts:17]: nope, check out the function `add_body_attribute` - I put it into utils 👍

File: src/instrumentations/amqplib/AmqplibInstrumentation.test.js (hunk 1 above)
Comment (saartochner) [reply src/instrumentations/amqplib/AmqplibInstrumentation.test.js]: it seems like they're taking tons of time, and I'm not sure what is their impact 🤷  please consult with Moses about it :)

File: src/instrumentations/amqplib/AmqplibInstrumentation.ts (hunk 2 above)
Comment (therightstuff) [reply src/instrumentations/amqplib/AmqplibInstrumentation.ts:17]: that's the python distro :)

File: src/instrumentations/amqplib/AmqplibInstrumentation.test.js (hunk 1 above)
Comment (therightstuff) [reply src/instrumentations/amqplib/AmqplibInstrumentation.test.js:15]: as discussed out of band, i'm considering this out of scope and have created https://lumigo.atlassian.net/browse/RD-11195
